text/00110-compatibility-strategy.md

```diff
@@ -0,0 +1,364 @@
+    run_tests
+}
+
+if [ ${CURRENT_VERSION} == "1.19.0" ]; then
+  
+    # version 1.20.0 of the framework bumped protocol version, so we expect an error message.
+    expect_error
+
+elif [ ${CURRENT_VERSION} == "1.20.0" ]; then
+
+    # version 1.21.0 of the framework bumped protocol version, so we expect an error message.
+    expect_error
+
+else
+
+    # common case, backwards compatible (i.e no protocol version bump)
+    # all old integration tests should pass.
+    expect_pass
+
+fi
+
+```
+
+In addition, we create a new file, which is responsible for validating the compatibility of new cli versions against old framework versions.
+
+### Things to note
+
+- Nothing really :)
+
+\+ **test-cli-compatibility.sh**
+
+```bash
+source ./common.bash
+
+# CLI should always support older framework versions, so these tests should always pass.
+run_integration_tests_perm ${PREVIOUS_VERSION} ${PREVIOUS_VERSION} ${CURRENT_VERSION}
+
+```
+
+## Developer Workflow
+
+Ideally, developers should mainly focus on code changes that are relevant to the functionality they are implementing, not necessarily concerning themselves with versioning issues. 
+
+This design attempts to provide a mechanism that, given the compatibility is broken, will guide the developer to the necessary code changes that need to be done. This way, developers don't have to *remember* anything, they simply won't be able to properly merge their changes until those versioning concerns are addressed.
+
+> Thats not to say we should stop thinking about compatibility when designing our changes, it simply alleviates the fear of unintentionally breaking compatibility.
+
+Following are the two possible scenarios that might break compatibility, with a diagram that represents the actions the developer would take in order for the change to be accepted. 
+
+### Introducing changes to the CLI
+
+This is really straight forward, new cli versions must always be compatible with older framework versions.
+
+1. Developer merges a PR to master.
+2. CLI compatibility tests are executed.
+3. They can result in one of two states:
+
+    - `pass`
+    - `fail`
+
+4. If tests pass, we are good. If they fail, fix the code and *GOTO* 1.
+
+<br>
+
+<!-- To edit: URL Decode -> Edit -> URL Encode -->
+
+<img src='https://g.gravizo.com/svg?digraph%20G%20%7B%0A%20%20%20%20aize%20%3D%224%2C4%22%3B%0A%20%20%20%20%22merge-to-master%22%3B%0A%20%20%20%20%22merge-to-master%22%20-%3E%20%22test-cli-compatibility.sh%22%3B%20%0A%20%20%20%20node%20%5Bshape%3Dbox%2Cstyle%3Dfilled%2Ccolor%3D%22yellowgreen%22%5D%3B%20%20%20%0A%20%20%20%20%22test-cli-compatibility.sh%22%20-%3E%20pass%3B%20%20%20%20%0A%20%20%20%20node%20%5Bshape%3Dbox%2Cstyle%3Dfilled%2Ccolor%3D%22crimson%22%5D%3B%20%20%20%0A%20%20%20%20%22test-cli-compatibility.sh%22%20-%3E%20fail%3B%20%20%20%20%0A%20%20%20%20fail%20-%3E%20%22merge-to-master%22%20%5Blabel%20%3D%20%22%20%20make-compatible%22%5D%0A%20%20%7D%0A'/>
+
+<br>
+
+### Introducing changes to the Framework
```

Comment:
This does not discuss the ergonomics of how a protocol change is introduced. I think most of our developer ergonomic mistakes are related to the fact that we actually have a protocol version. Ideally we should check for compatibility at the feature level and not at the protocol level.

For example, when we introduced support for custom docker file name, the code at the framework level could have verified that this feature is supported by the CLI (i.e. the CLI sends some sort of feature map in the context) and will only fail if users wish to use this feature with a CLI that doesn't support it.

Reply:
I intentionally didn't attempt to fix the ergonomics of the protocol. I tried to propose a validation mechanism that validates our compatibility promises, regardless of the underlying implementation. This is described in the `Motivation` section. 

I agree that we can be more granular in our compatibility checks, but I wonder if its worth it. This feels like a very complex testing matrix, and perhaps a better approach is simply making sure we require an upgrade regardless of if the feature is actually used or not, at least for now. 

*How do we make sure we require an upgrade? See https://github.com/aws/aws-cdk-rfcs/pull/111#discussion_r374326231*

In any case, a pre-requisite to starting this work is to re-iterate and align our compatibility promises themselves, as this affects everything.